### PR TITLE
[Brownfield] Settings manager

### PR DESCRIPTION
**Pull request by annakocheshkova**
Fully rewrote settings manager and 100% covered it with tests.

### CODE REVIEW

File: android/app/src/main/java/com/microsoft/codepush/common/datacontracts/CodePushPendingUpdate.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
+package com.microsoft.codepush.common.datacontracts;
+
+import com.google.gson.annotations.SerializedName;
+
+/**
+ * Created by anna.kocheshkova on 2/2/2018.
+ */
+
+public class CodePushPendingUpdate {
+
```
Comment (ruslan-bikkinin) [inline android/app/src/main/java/com/microsoft/codepush/common/datacontracts/CodePushPendingUpdate.java]: Empty line.

File: android/app/src/main/java/com/microsoft/codepush/common/datacontracts/CodePushPendingUpdate.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
+package com.microsoft.codepush.common.datacontracts;
+
+import com.google.gson.annotations.SerializedName;
+
+/**
+ * Created by anna.kocheshkova on 2/2/2018.
```
Comment (ruslan-bikkinin) [inline android/app/src/main/java/com/microsoft/codepush/common/datacontracts/CodePushPendingUpdate.java]: Redundant comment header.

File: android/app/src/main/java/com/microsoft/codepush/common/CodePushConstants.java [hunk 3]
```diff
@@ -5,6 +5,7 @@
  */
 public final class CodePushConstants {
 
+
```
Comment (ruslan-bikkinin) [inline android/app/src/main/java/com/microsoft/codepush/common/CodePushConstants.java]: Empty line.

File: android/app/src/main/java/com/microsoft/codepush/common/managers/SettingsManager.java [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+     */
+    private SharedPreferences mSettings;
+
+    /**
+     * Creates an instance of {@link SettingsManager} with {@link Context} provided.
+     *
+     * @param applicationContext current application context.
+     */
+    public SettingsManager(Context applicationContext) {
+        mSettings = applicationContext.getSharedPreferences(CodePushConstants.CODE_PUSH_PREFERENCES, 0);
```
[the remark this answers is not in the document]
Comment (annakocheshkova) [reply android/app/src/main/java/com/microsoft/codepush/common/managers/SettingsManager.java]: @ruslan-bikkinin Didn't quite get the point. What do you mean by part of the Context?

File: android/app/src/main/java/com/microsoft/codepush/common/utils/CodePushUtils.java [hunk 5]
```diff
@@ -189,7 +190,7 @@ public static String convertObjectToJsonString(Object object) {
      * @return instance of T.
      */
     @SuppressWarnings("WeakerAccess")
-    public static <T> T convertStringToObject(String stringObject, Class<T> classOfT) {
+    public static <T> T convertStringToObject(String stringObject, Class<T> classOfT) throws JsonSyntaxException {
```
[the remark this answers is not in the document]
Comment (annakocheshkova) [reply android/app/src/main/java/com/microsoft/codepush/common/utils/CodePushUtils.java]: @ruslan-bikkinin Well, `Gson` method does.

File: android/app/src/main/java/com/microsoft/codepush/common/managers/SettingsManager.java [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+        }
+    }
+
+    /**
+     * Checks whether an update with the following hash has failed.
+     *
+     * @param packageHash hash to check.
+     * @return <code>true</code> if there is a failed update with provided hash.
+     */
+    public boolean isFailedHash(String packageHash) {
```
[the remark this answers is not in the document]
Comment (annakocheshkova) [reply android/app/src/main/java/com/microsoft/codepush/common/managers/SettingsManager.java]: @ruslan-bikkinin Sure, this sounds good.

File: android/app/src/main/java/com/microsoft/codepush/common/managers/SettingsManager.java [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+     */
+    public CodePushPendingUpdate getPendingUpdate() {
+        String pendingUpdateString = mSettings.getString(PENDING_UPDATE_KEY, null);
+        if (pendingUpdateString == null) {
+            return null;
+        }
+        try {
+            return CodePushUtils.convertStringToObject(pendingUpdateString, CodePushPendingUpdate.class);
+        } catch (JsonSyntaxException e) {
+            AppCenterLog.logAssert(LOG_TAG, "Unable to parse pending update metadata " + pendingUpdateString +
```
[the remark this answers is not in the document]
Comment (annakocheshkova) [reply android/app/src/main/java/com/microsoft/codepush/common/managers/SettingsManager.java]: @ruslan-bikkinin Have replaced logAssert with error. Don't think this case is big enough for throwing an exception.

File: android/app/src/main/java/com/microsoft/codepush/common/managers/SettingsManager.java [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+        String failedUpdatesString = mSettings.getString(FAILED_UPDATES_KEY, null);
+        if (failedUpdatesString == null) {
+            return new ArrayList<>();
+        }
+        try {
+            return new ArrayList<>(Arrays.asList(CodePushUtils.convertStringToObject(failedUpdatesString, CodePushLocalPackage[].class)));
+        } catch (JsonSyntaxException e) {
+
+            /* Unrecognized data format, clear and replace with expected format. */
+            List<CodePushLocalPackage> emptyArray = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (annakocheshkova) [reply android/app/src/main/java/com/microsoft/codepush/common/managers/SettingsManager.java]: Have added `AppCenterLog.error()`. Don't think this case is big enough for throwing an exception.

File: android/app/src/main/java/com/microsoft/codepush/common/managers/SettingsManager.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (ruslan-bikkinin) [reply android/app/src/main/java/com/microsoft/codepush/common/managers/SettingsManager.java]: @annakocheshkova I meant, that we just call `applicationContext.getSharedPreferences` and nothing more is needed from `applicationContext` at all.

[the remark this answers is not in the document]
Comment (annakocheshkova) [reply android/app/src/main/java/com/microsoft/codepush/common/managers/SettingsManager.java]: @ruslan-bikkinin Do you suggest passing not the Context, but shared preferences? This is not a good practice to do that, AFAIK.  Every settings manager I've seen is initialized that way.

[the remark this answers is not in the document]
Comment (ruslan-bikkinin) [reply android/app/src/main/java/com/microsoft/codepush/common/managers/SettingsManager.java]: @annakocheshkova gotcha, just ignore this comment branch then.

File: android/app/build.gradle [hunk 9]
```diff
@@ -7,7 +7,7 @@ jacoco {
 
 android {
     compileSdkVersion 23
-    buildToolsVersion "23.0.1"
+    buildToolsVersion '23.0.1'
```
Comment (ruslan-bikkinin) [inline android/app/build.gradle]: Please remove this unnecessary change.
